### PR TITLE
[2103]  Add support for edge reconnection

### LINKED ISSUE #2103: Add support for edge reconnection
See https://reactflow.dev/docs/examples/edges/updatable-edge/

### PR DESCRIPTION
**Pull request by mcharfadi**
Bug : https://github.com/eclipse-sirius/sirius-components/issues/2103

### CODE REVIEW

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx [hunk 1]
```diff
@@ -11,9 +11,9 @@
  *     Obeo - initial API and implementation
  *******************************************************************************/
 
-import { Selection, SelectionEntry } from '@eclipse-sirius/sirius-components-core';
+import { Selection, SelectionEntry, useMultiToast } from '@eclipse-sirius/sirius-components-core';
```
Comment (sbegaudeau) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx]: In a similar fashion as my comment on hide, fade, you could prevent the addition of additional complexity in DiagramRender using an API like this:

```
import { useReconnectEdge } from './reconnect-edge/useReconnectEdge';

// ...

const { reconnectEdge } = useReconnectEdge();
```

**Review by sbegaudeau — DISMISSED** [review]
Given the improvement mades in the pull request to add the support for direct edit, we will not add additional complexity to the DiagramRenderer. Especially GraphQL mutations etc.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx [hunk 2]
```diff
@@ -151,6 +155,10 @@ export const DiagramRenderer = ({ diagram, selection, setSelection }: DiagramRen
     });
   };
 
+  const onEdgeUpdate = (oldEdge, newConnection) => {
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx]: oldEdge and newConnection should be typed

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/reconnect-edge/useReconnectEdge.types.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+export interface GQLReconnectEdgePayload {
+  __typename: string;
+}
+
+export enum GQLReconnectKind {
+  SOURCE = 'SOURCE',
+  TARGET = 'TARGET',
+}
+
+export interface GQLSuccessPayload extends GQLReconnectEdgePayload {
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/reconnect-edge/useReconnectEdge.types.ts]: Remove the success payload since it is not used

File: packages/diagrams/frontend/sirius-components-diagrams/src/representation/DiagramRepresentation.tsx [hunk 4]
```diff
@@ -21,8 +21,8 @@ import {
   useMultiToast,
 } from '@eclipse-sirius/sirius-components-core';
 import { SelectionDialog } from '@eclipse-sirius/sirius-components-selection';
-import makeStyles from '@material-ui/core/styles/makeStyles';
 import Typography from '@material-ui/core/Typography';
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams/src/representation/DiagramRepresentation.tsx]: Why is this file modified?

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx [hunk 5]
```diff
@@ -151,6 +155,10 @@ export const DiagramRenderer = ({ diagram, selection, setSelection }: DiagramRen
...
+    reconnectEdge(oldEdge, newConnection);
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx]: Be careful, a react flow Connection is not a react flow Edge.
You may need to update the `reconnectEdge` api so support Connection instead of Edge for the second argument

**Review by gcoutable — CHANGES_REQUESTED** [review]

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx [hunk 6]
```diff
@@ -179,6 +189,7 @@ export const DiagramRenderer = ({ diagram, selection, setSelection }: DiagramRen
       edgeTypes={edgeTypes}
       onKeyDown={onKeyDown}
       onEdgesChange={handleEdgesChange}
+      onEdgeUpdate={onEdgeUpdate}
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx]: You can directly use _reconnectEdge_ here

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/reconnect-edge/useReconnectEdge.tsx [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    }
+  }
+`;
+
+export const useReconnectEdge = (): UseReconnectEdge => {
+  const { addErrorMessage } = useMultiToast();
+  const { diagramId, editingContextId } = useContext<DiagramContextValue>(DiagramContext);
+
+  const isErrorPayload = (payload: GQLReconnectEdgePayload): payload is GQLErrorPayload =>
+    payload.__typename === 'ErrorPayload';
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/reconnect-edge/useReconnectEdge.tsx]: Move this outside of the hook

**Review by gcoutable — CHANGES_REQUESTED** [review]

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx (hunk 6 above)
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramRenderer.tsx]: done

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/reconnect-edge/useReconnectEdge.tsx (hunk 7 above)
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/reconnect-edge/useReconnectEdge.tsx]: done

**Review by gcoutable — APPROVED** [review]
